qt/aqt/progress.py

```diff
@@ -223,25 +224,49 @@ def update(
             self._win.form.progressBar.setValue(self._counter)
 
     def finish(self) -> None:
-        # we must latch the levels update and perform it after cleanup has occured or we expose ourselves to a race
-        # condition where a second progress could see levels == 0 and wrongly assume everything is in a clean state
-        next_levels = self._levels - 1
-        next_levels = max(0, next_levels)
-        if next_levels == 0:
-            if self._win:
-                self._closeWin()
-            if self._busy_cursor_timer:
-                self._busy_cursor_timer.stop()
-                self._busy_cursor_timer = None
-            self._restore_cursor()
-            if self._show_timer:
-                self._show_timer.stop()
-                self._show_timer = None
-        if self._backend_timer:
-            self._backend_timer.stop()
-            self._backend_timer.deleteLater()
-            self._backend_timer = None
-        self._levels = next_levels
+        def do_window_cleanup(future: Future | None = None):
+            # this method can be called in an async fashion from taskman where a future
+            # is passed or in synchronous manner from the main thread
+            if future is not None:
+                future.result()
+
+            next_levels = self._levels - 1
+            next_levels = max(0, next_levels)
+            if next_levels == 0:
+                if self._win:
+                    self._closeWin()
+                if self._busy_cursor_timer:
+                    self._busy_cursor_timer.stop()
+                    self._busy_cursor_timer = None
+                self._restore_cursor()
+                if self._show_timer:
+                    self._show_timer.stop()
+                    self._show_timer = None
+            if self._backend_timer:
+                self._backend_timer.stop()
+                self._backend_timer.deleteLater()
+                self._backend_timer = None
+            self._levels = next_levels
+
+        # if the window is not currently shown, we can do cleanup immediately, if it is
+        # currently shown then we need to give the window system a half-second to
+        # present the window before we close it again - fixes progress window getting
+        # stuck, especially on ubuntu 16.10+
+        # NOTE: we must not yield control if the window is not shown since we don't want
+        # to expose ourselves to the possibility of something showing the window in the
+        # meantime
+        if self._shown:
+            elapsed_time = time.monotonic() - self._shown
+            if (time_to_wait := 0.5 - elapsed_time) > 0:
+                self.mw.taskman.run_in_background(
```

Comment:
Using sleep() in a background thread could be done more simply with self.single_shot()

Reply:
Oh nice that is perfect!

Reply:
@dae I just looked at this a little more and we can't actually use `single_shot` here since it wraps the callback and won't call it until `_levels` is 0 which is the whole point of our callback :/ I'll stick to the threadpool for now and add a comment about it.

Reply:
Sorry to lead you down the wrong path there -  I'd forgotten about that check.